<a href="https://colab.research.google.com/github/Aditya112008/Class139DemograhicFiltering/blob/main/Class_139DemographicFiltering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install kaggle

In [2]:
from google.colab import files
files.upload()

Saving kaggle.json to kaggle.json


{'kaggle.json': b'{"username":"adityagokarn","key":"1b01cb0b939be3122292f1da1f3bbb27"}'}

In [3]:
!mkdir -p ~/.kaggle
!cp kaggle.json ~/.kaggle/

!chmod 600 ~/.kaggle/kaggle.json

In [4]:
!kaggle datasets download -d tmdb/tmdb-movie-metadata

  0% 0.00/8.89M [00:00<?, ?B/s]
100% 8.89M/8.89M [00:00<00:00, 82.0MB/s]


In [5]:
!ls

kaggle.json  sample_data  tmdb-movie-metadata.zip


In [6]:
!unzip tmdb-movie-metadata.zip

Archive:  tmdb-movie-metadata.zip
  inflating: tmdb_5000_credits.csv   
  inflating: tmdb_5000_movies.csv    


In [7]:
!ls

kaggle.json  tmdb_5000_credits.csv  tmdb-movie-metadata.zip
sample_data  tmdb_5000_movies.csv


In [8]:
import pandas as pd 
import numpy as np 

df1=pd.read_csv('tmdb_5000_credits.csv')
df2=pd.read_csv('tmdb_5000_movies.csv')

In [9]:
df1.head()

movie_id  ...                                               crew
0     19995  ...  [{"credit_id": "52fe48009251416c750aca23", "de...
1       285  ...  [{"credit_id": "52fe4232c3a36847f800b579", "de...
2    206647  ...  [{"credit_id": "54805967c3a36829b5002c41", "de...
3     49026  ...  [{"credit_id": "52fe4781c3a36847f81398c3", "de...
4     49529  ...  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

[5 rows x 4 columns]

In [10]:
df2.head()

budget  ... vote_count
0  237000000  ...      11800
1  300000000  ...       4500
2  245000000  ...       4466
3  250000000  ...       9106
4  260000000  ...       2124

[5 rows x 20 columns]

In [11]:
df1.columns = ['id','tittle','cast','crew']
df2= df2.merge(df1,on='id')

df2.head(5)

budget  ...                                               crew
0  237000000  ...  [{"credit_id": "52fe48009251416c750aca23", "de...
1  300000000  ...  [{"credit_id": "52fe4232c3a36847f800b579", "de...
2  245000000  ...  [{"credit_id": "54805967c3a36829b5002c41", "de...
3  250000000  ...  [{"credit_id": "52fe4781c3a36847f81398c3", "de...
4  260000000  ...  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

[5 rows x 23 columns]

# Demographic Filtering

We need a metric or score to rate the movie,
We need to calculate the score for every movie and then we finally need to sort the scores and recommend the best rated movie to the user

**imdb(internet movie database) has created a formula it is known as weighted rating and it is famously used in the industry for the same thing (to rate the movie) to get a score for their products or items**

Weighted Rating (WR) = ((v/v+m)*R) + ((m/v+m)*C)
v - it is the number of votes for a particular movie 

m - is the number of votes required to be listed in the chat 

R - average rating of the movie
 
C - mean votes across the whole report 

In [12]:
C = df2['vote_average'].mean()
print(C)

6.092171559442011


In [13]:
m = df2['vote_count'].quantile(0.9)
print(m)

1838.4000000000015


For a movie to feature in the chats it must have more votes than atleast 90% of the rest of the movies in the list 

a value more than 90% of the movies in terms of vote count.

Let's create a new data frame with all the movies that have more votes than the value of "m"



In [14]:
q_movies = df2.copy().loc[df2['vote_count'] >= m]
print(q_movies.shape)

(481, 23)


copy function copies the contents of df2 
we are applying a condition on the copied content with the loc function (we are saying that the vote_count of the movie shall be equal to or greater than "m"

We are storing this new data frame into a variable named as q_movies 

We can see that there are 481 movies which qualify to be in the list 

In [15]:
def weighted_rating(x, m=m, C=C):
    v = x['vote_count']
    R = x['vote_average']
    return (v/(v+m) * R) + (m/(m+v) * C)

q_movies['score'] = q_movies.apply(weighted_rating, axis=1)

We will define a function named as weighted_rating() to create a new column "score" in all the rows. 

we are passing individual row, "m" and "C" values to the function. 

storing the vote count in the "v",

storing vote_average in variable "R"

we have used apply() function to use the formula because we are dealing with a data frame, and we need to pass individual rows to our function. apply() function takes care of that 

storing the output in "score" column in the data frame 


In [16]:
q_movies = q_movies.sort_values('score', ascending=False)
q_movies[['title', 'vote_count', 'vote_average', 'score']].head(10)

title  ...     score
1881                       The Shawshank Redemption  ...  8.059258
662                                      Fight Club  ...  7.939256
65                                  The Dark Knight  ...  7.920020
3232                                   Pulp Fiction  ...  7.904645
96                                        Inception  ...  7.863239
3337                                  The Godfather  ...  7.851236
95                                     Interstellar  ...  7.809479
809                                    Forrest Gump  ...  7.803188
329   The Lord of the Rings: The Return of the King  ...  7.727243
1990                        The Empire Strikes Back  ...  7.697884

[10 rows x 4 columns]

We have to sort the data frame on the "score" column. 
sort_values() function of dataframe sorts the values of "score" column

asecending=False - creates a descending order 

we will also output the title, vote_count, vote_average, score (weighted rating)  of top 10 movies.

the function head(10) will show the first top 10 rows from the data frame


In [17]:
import plotly.express as px

fig = px.bar((q_movies.head(10).sort_values('score', ascending=True)), x="score", y="title", orientation='h')
fig.show()

In [18]:
import plotly.express as px

fig = px.bar((q_movies.head(10).sort_values('score', ascending=False)), x="score", y="title", orientation='h')
fig.show()

Demographic Recommendation provides a general chart of recommended movies to all the users

they are not sensitive to the interests of a particular user 

this is where a more refined system - content based filtering comes into play 
